presto-main/src/main/java/io/prestosql/operator/PageUtils.java

```diff
@@ -15,45 +15,57 @@
 
 import io.prestosql.spi.Page;
 import io.prestosql.spi.block.Block;
-import io.prestosql.spi.block.LazyBlock;
 
+import java.util.function.Consumer;
 import java.util.function.LongConsumer;
 
+import static io.prestosql.spi.block.LazyBlock.listenForLoads;
+import static java.util.Objects.requireNonNull;
+
 public final class PageUtils
 {
     private PageUtils() {}
 
-    public static Page recordMaterializedBytes(Page page, LongConsumer sizeInBytesConsumer)
+    public static void recordMaterializedBytes(Page page, LongConsumer sizeInBytesConsumer)
     {
         // account processed bytes from lazy blocks only when they are loaded
-        Block[] blocks = new Block[page.getChannelCount()];
         long loadedBlocksSizeInBytes = 0;
-        boolean allBlocksLoaded = true;
 
         for (int i = 0; i < page.getChannelCount(); ++i) {
             Block block = page.getBlock(i);
-            if (block.isLoaded()) {
-                loadedBlocksSizeInBytes += block.getSizeInBytes();
-                blocks[i] = block;
-            }
-            else {
-                blocks[i] = new LazyBlock(page.getPositionCount(), () -> {
-                    Block loadedBlock = block.getLoadedBlock();
-                    sizeInBytesConsumer.accept(loadedBlock.getSizeInBytes());
-                    return loadedBlock;
-                });
-                allBlocksLoaded = false;
-            }
+            long initialSize = block.getSizeInBytes();
+            loadedBlocksSizeInBytes += initialSize;
+            listenForLoads(block, new BlockSizeListener(block, sizeInBytesConsumer, initialSize));
```

Comment:
I think this might have performance impact when blocks are loaded and tiny. It would be great to re-run `BenchmarkScanFilterAndProjectOperator`